### PR TITLE
added some new warnings DJ09, DJ10, DJ11

### PR DESCRIPTION
**Pull request by avallbona**
added some new warnings DJ09, DJ10, DJ11

- Added `DJ09` - Model must define `class Meta`
- Added `DJ10` - Class Meta from Model has to define `verbose_name`
- Added `DJ11` - Class Meta from Model has to define `verbose_name_plural`
- Fixed some flake8 issues in package and tests and added pypi badge
- Refactored some tests
- Bumped version number

### CODE REVIEW

**Comment by rocioar** [comment]
@avallbona I'm not sure if `verbose_name` and `verbose_name_plural` should be required, most of the times the names of the models work just fine, for example: `Artist` and `Artists`. You don't really need to set those on such cases.

**Comment by avallbona** [comment]
Hello, it is not mandatory but I think it is a good practice because it allows us, more easily, to introduce translatable strings for the name of our models. The vast majority of the projects in which I participate are multi-language. In the same way that we add a verbose_name to the fields of the model, I think it is also consistent to add the name of the models in both singular and plural forms. 

As a counterexample, there are also many cases of plural forms that do not end in "s" in English:

- woman -> women
- man -> men
- person -> people

**Comment by avallbona** [comment]
Hi @rocioar, given that you can always disable the warnings via the exclude option, in case you don't need it, will you reconsider merging the changes? Kind regards.

**Comment by rocioar** [comment]
@avallbona given that you can configure if you want the check or not, I think I will be ok with this changes. I will do a review of the code soon.

One thing that I noticed is that this PR is bumping the version, we shouldn't do this on this PR, the way we usually do it is in a single PR that just takes care of bumping the version, I can do that after this PR is merged.

**Comment by avallbona** [comment]
hi @rocioar, I've left the version as 0.0.4, like it was before.
Kind regards.

File: flake8_django/checker.py [hunk 1]
```diff
@@ -1,8 +1,9 @@
 import ast
 
-from flake8_django.checkers import ModelDunderStrMissingChecker, ModelFieldChecker, ModelFormChecker, URLChecker, RenderChecker
+from flake8_django.checkers import ModelDunderStrMissingChecker, ModelFieldChecker, ModelFormChecker, \
+    URLChecker, RenderChecker, ModelMetaChecker
 
-__version__ = '0.0.3'
+__version__ = '0.0.4'
```
Comment (rocioar) [inline flake8_django/checker.py]: good catch 👍

File: flake8_django/checkers/model_meta.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    description = 'Model: {model} has no verbose_name_plural'
+
+
+class ModelMetaChecker(BaseModelChecker):
+    model_name_lookup = 'Model'
+
+    def checker_applies(self, node):
+        for base in node.bases:
+            if self.is_model_name_lookup(base) or self.is_models_name_lookup_attribute(base):
+                return True
```
Comment (rocioar) [inline flake8_django/checkers/model_meta.py]: @avallbona could you add a check here for abstract model, if it's an abstract model we shouldn't be throwing an error. There is `is_abstract_model` on `BaseModelChecker` that you can use for that purpose.

File: flake8_django/checkers/model_meta.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    def has_verbose_name_plural(self, meta_class_node):
+        return self._has_element(meta_class_node, 'verbose_name_plural')
+
+    @staticmethod
+    def _has_element(element, target_name):
+        for node in ast.iter_child_nodes(element):
+            if not isinstance(node, ast.Assign):
+                continue
+            attr = node.targets[0].id
```
Comment (rocioar) [inline flake8_django/checkers/model_meta.py]: Could you add a check here for `isinstance(node.targets[0], ast.Name)`. This is to avoid the issue in which we could have something like this:

```python
class Meta:
    a, b = 1
```

File: tests/fixtures/model_form_fields_all.py [hunk 4]
```diff
@@ -13,4 +13,4 @@ class User4(ModelForm):
 
     class Meta:
         model = User2
-        fields = b'__all__'
\ No newline at end of file
+        fields = '__all__'
```
Comment (rocioar) [inline tests/fixtures/model_form_fields_all.py]: @avallbona please leave this as it was since this is to cover a case in which the attribute could be a bytes literal, and we don't want the checkers to fail in this case.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,6 +1,18 @@
...
+------------------
+
+**Improvements**
+
+- Added `DJ09` - Model must define `class Meta`
+- Added `DJ10` - Class Meta from Model has to define `verbose_name`
+- Added `DJ11` - Class Meta from Model has to define `verbose_name_plural`
+- Fixed some flake8 issues in package and tests and added pypi badge
+- Refactored some tests
+- Bumped version number
```
[the remark this answers is not in the document]
Comment (avallbona) [reply CHANGELOG.md]: done

File: README.md [hunk 6]
```diff
@@ -1,5 +1,6 @@
 # flake8-django
 
+[![pypi](https://img.shields.io/pypi/v/flake8-django.svg)](https://pypi.python.org/pypi/flake8-django/)
```
[the remark this answers is not in the document]
Comment (avallbona) [reply README.md]: done

File: flake8_django/checker.py [hunk 7]
```diff
@@ -1,8 +1,9 @@
 import ast
 
-from flake8_django.checkers import ModelDunderStrMissingChecker, ModelFieldChecker, ModelFormChecker, URLChecker, RenderChecker
+from flake8_django.checkers import ModelDunderStrMissingChecker, ModelFieldChecker, ModelFormChecker, \
+    URLChecker, RenderChecker, ModelMetaChecker
```
[the remark this answers is not in the document]
Comment (avallbona) [reply flake8_django/checker.py]: done

File: flake8_django/checkers/model_meta.py (hunk 3 above)
Comment (avallbona) [reply flake8_django/checkers/model_meta.py]: done

File: tests/fixtures/model_form_fields_all.py (hunk 4 above)
Comment (avallbona) [reply tests/fixtures/model_form_fields_all.py]: done

File: flake8_django/checkers/model_meta.py (hunk 2 above)
Comment (avallbona) [reply flake8_django/checkers/model_meta.py]: done

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (rocioar) [reply README.md]: Could you remove the pypi badge from here now? Thanks!

File: tests/conftest.py [hunk 8]
```diff
@@ -0,0 +1,3 @@
+
+
+
```
Comment (rocioar) [inline tests/conftest.py]: @avallbona I don't think we need to add this file, unless we were using it. Could you remove it?

File: tests/utils.py [hunk 9]
```diff
@@ -17,3 +17,17 @@ def load_fixture_file(filename):
...
+    assert that a specific error_code is into the results
+
+    :param error_code:
+    :param result:
+    :return:
+    """
+    for element in result:
+        if error_code in element[2]:
+            return True
+    return False
```
Comment (rocioar) [inline tests/utils.py]: @avallbona could you add a new line at the end of the file? Thanks!

```suggestion
    return False

```

File: tests/test_urls.py [hunk 10]
```diff
@@ -41,4 +42,5 @@ def test_url_include_without_namespace_fails(code):
     "re_path('/test/', includes('test2', namespace='test-2'))",
 ])
 def test_url_include_with_namespace_success(code):
-    assert len(run_check(code)) == 0
+    result = run_check(code)
+    assert not error_code_in_result('DJ05', result)
```
Comment (rocioar) [inline tests/test_urls.py]: could you add a new line at the end of the file? Thanks!

```suggestion
    assert not error_code_in_result('DJ05', result)

```

File: flake8_django/checkers/base_model_checker.py [hunk 11]
```diff
@@ -10,11 +10,12 @@ class BaseModelChecker(Checker):
...
     model_name_lookup = None
 
-    def _is_abstract_and_set_to_true(self, element):
+    @staticmethod
+    def _is_abstract_and_set_to_true(element):
         return (
                 isinstance(element, ast.Assign)
-                and element.value.value is True
                 and any(target.id == 'abstract' for target in element.targets)
+                and element.value.value is True
```
Comment (rocioar) [inline flake8_django/checkers/base_model_checker.py]: @avallbona wondering, why did you change the order of the checks? I think we could check for value and avoid doing that for through the target elements first, but maybe there is a reason for this change.

File: CHANGELOG.md [hunk 12]
```diff
@@ -7,6 +7,11 @@ Change Log
 **Improvements**
 
 - Changed `DJ08` check for `__str__` method to ignore abstract models (@denizdogan)
+- Added `DJ09` - Model must define `class Meta`
+- Added `DJ10` - Class Meta from Model has to define `verbose_name`
+- Added `DJ11` - Class Meta from Model has to define `verbose_name_plural`
+- Fixed some flake8 issues in package and tests and added pypi badge
```
Comment (rocioar) [inline CHANGELOG.md]: Could you update this line to include your name, and also exclude "and added pypi badge" since this has been added on another PR.

```suggestion
- Fixed some flake8 issues in test files (@avallbona)
```

Also please include your username in all the other lines. Thanks

File: .gitignore [hunk 13]
```diff
@@ -105,3 +105,4 @@ venv.bak/
 
 # IntelliJ
 .idea
+Pipfile*
```
Comment (rocioar) [inline .gitignore]: @avallbona could you add a line at the end of the file?

```suggestion
Pipfile*

```

File: tests/conftest.py (hunk 8 above)
Comment (avallbona) [reply tests/conftest.py]: done

File: tests/utils.py (hunk 9 above)
Comment (avallbona) [reply tests/utils.py]: done

File: tests/test_urls.py (hunk 10 above)
Comment (avallbona) [reply tests/test_urls.py]: done

File: CHANGELOG.md (hunk 12 above)
Comment (avallbona) [reply CHANGELOG.md]: done

File: .gitignore (hunk 13 above)
Comment (avallbona) [reply .gitignore]: done

File: flake8_django/checkers/base_model_checker.py (hunk 11 above)
Comment (avallbona) [reply flake8_django/checkers/base_model_checker.py]: @rocioar when you have a model that does not have the abstract attribute and instead has a "verbose_name" attribute when you check if "element.value.value is True" the "element.value" is a string and does not have the attribute value, then if fails saying:

`AttributeError: 'Str' object has no attribute 'value'`

For this I've changed the order, first check if we have an attribute called "abstract" and only if we have it check if its value is True.

Comment (rocioar) [reply flake8_django/checkers/base_model_checker.py]: Oh I see good catch, let's check for NameConstant then first:

This should work:

```python
return (
    isinstance(element, ast.Assign)
    and any(target.id == 'abstract' for target in element.targets)
    and isinstance(element.value, ast.NameConstant)
    and element.value.value is True
)
```

File: CHANGELOG.md (hunk 12 above)
Comment (rocioar) [reply CHANGELOG.md]: @avallbona the line didn't change at all, maybe you forgot to push the commit?

File: flake8_django/checkers/model_meta.py (hunk 3 above)
Comment (avallbona) [reply flake8_django/checkers/model_meta.py]: Yes, you're right, when the first there is the tuple assign as you point it would fail. 
Added the theck.

File: flake8_django/checkers/base_model_checker.py (hunk 11 above)
Comment (avallbona) [reply flake8_django/checkers/base_model_checker.py]: done. And added an extra check into the iteration in order to avoid the case when, in the meta class you have a tuple assignation, as you pointed before.

File: CHANGELOG.md (hunk 12 above)
Comment (avallbona) [reply CHANGELOG.md]: I was still working on the issue, sorry. I've just pushed the commit now.

File: flake8_django/checkers/base_model_checker.py (hunk 11 above)
Comment (rocioar) [reply flake8_django/checkers/base_model_checker.py]: I think you will also have to fix the identantion, it's 8 characters now.

File: CHANGELOG.md [hunk 14]
```diff
@@ -7,6 +7,11 @@ Change Log
 **Improvements**
 
 - Changed `DJ08` check for `__str__` method to ignore abstract models (@denizdogan)
+- Added `DJ09` - Model must define `class Meta`
+- Added `DJ10` - Class Meta from Model has to define `verbose_name`
+- Added `DJ11` - Class Meta from Model has to define `verbose_name_plural`
+- Fixed some flake8 issues in test files (@avallbona)
+- Refactored some tests
```
Comment (rocioar) [inline CHANGELOG.md]: @avallbona you also have to add your name to all the other lines that have changes made by you

Comment (avallbona) [reply CHANGELOG.md]: done

**Comment by washeck** [comment]
I personally find the checks introduced in this PR to be too strict. Many Django projects can live without using verbose_name especially when Django is used as a backend-only solution with JS frontend.

If it is possible, I would find it good to include these checks but make them disabled by default. My fear is that having too strict defaults will reduced the chance of the plugin to get adopted by developers.

**Comment by avallbona** [comment]
> I personally find the checks introduced in this PR to be too strict. Many Django projects can live without using verbose_name especially when Django is used as a backend-only solution with JS frontend.
> 
> If it is possible, I would find it good to include these checks but make them disabled by default. My fear is that having too strict defaults will reduced the chance of the plugin to get adopted by developers.

Hi there, you can always disable these checks via the ignore parameter. E.g.:

`ignore=DJ09,DJ10,DJ11`

it's like you can do it with max-line-length parameter

`max-line-length=120`

On the other hand, @rocioar when do you think changes would be merged into master?

Thanks and kind regards. 
Andreu

**Comment by blueyed** [comment]
JFI: flake8 supports `--extend-ignore` by now, which allows to keep default ignores.

**Comment by rocioar** [comment]
@avallbona It's almost good to go! I made a comment regarding identation here: https://github.com/rocioar/flake8-django/pull/35/files#r284003395

Also, would you mind removing the `error_code_in_result` utility?, and just check with `'DJ05' in result`, or `'DJ05' not in result`. This is simpler, and also easy to read.

Once that is fixed, I will merge the PR, I just tested it throughly and everything works as expected.

Thank you!

**Comment by avallbona** [comment]
> @avallbona It's almost good to go! I made a comment regarding identation here: https://github.com/rocioar/flake8-django/pull/35/files#r284003395
> 
> Also, would you mind removing the `error_code_in_result` utility?, and just check with `'DJ05' in result`, or `'DJ05' not in result`. This is simpler, and also easy to read.
> 
> Once that is fixed, I will merge the PR, I just tested it throughly and everything works as expected.
> 
> Thank you!

Hi @rocioar, I'm sorry but I don't see the comment about the indentation.

On the other hand, I think that `error_code_in_result` function it's necessary because we're searching the error code inside a list of tuples. It can return more than one error. Beside this, I think, if someday, the checker changes the way it returns the errors it would be simpler to change only inside the utility function than changes all the tests code.

Kind regards.

File: flake8_django/checkers/base_model_checker.py (hunk 11 above)
Comment (rocioar) [reply flake8_django/checkers/base_model_checker.py]: @avallbona you were probably not seeing this comment because it was marked as resolved, please check identation on this lines.

**Comment by rocioar** [comment]
@avallbona You are right, sorry about that, I didn’t properly read it.

Regarding the lines with wrong identation, I’ve marked the comment as unresolved, you should be able to see it now.

File: flake8_django/checkers/base_model_checker.py (hunk 11 above)
Comment (avallbona) [reply flake8_django/checkers/base_model_checker.py]: @rocioar fixed the indentation issue.
Thanks for the review.

**Comment by GitRon** [comment]
@avallbona Any update on the PR? I think your changes do make sense, would like to have them in the code.

**Comment by avallbona** [comment]
Hi @GitRon, no news, sorry.
Hopefully some day changes will be merged.
Kind regards

**Comment by rocioar** [comment]
@avallbona Hey, on my last review I requested if you could check on a comment you missed: https://github.com/rocioar/flake8-django/pull/35/files#r284003395.

Once that's done, I will merge the PR.

File: flake8_django/checkers/model_meta.py (hunk 3 above)
Comment (avallbona) [reply flake8_django/checkers/model_meta.py]: Done in tests/fixtures/model_with_meta.py and tests/test_model_meta.py

**Comment by rocioar** [comment]
@avallbona Thanks a lot for your contribution to the project! I was actually referring to a comment regarding indentation, which I saw you had already replied back in October, I didn't see that you had replied at that time, one way to ensure the person sees the changes is by re-requesting a review.

I will soon release a new version of the package with the changes that we have on master :)

**Comment by GitRon** [comment]
@rocioar awesome!
